LLama/LLamaContext.cs

```diff
@@ -230,6 +233,64 @@ public void SaveState(string filename, LLamaSeqId sequence)
                 fileStream.SetLength(writtenBytes);
         }
 
+        /// <summary>
+        /// Save the state of a particular sequence to specified path. Also save some extra data which will be returned when loading.
+        /// Data saved with this method <b>must</b> be saved with <see cref="LoadState(string, LLamaSeqId, out byte[])"/>
+        /// </summary>
+        /// <param name="filename"></param>
+        /// <param name="sequence"></param>
+        /// <param name="header"></param>
+        internal void SaveState(string filename, LLamaSeqId sequence, ReadOnlySpan<byte> header)
+        {
+            // Delete that file before overwriting it
+            if (File.Exists(filename))
+                File.Delete(filename);
+
+            // Estimate size of state to write to disk, this is always equal to or greater than the actual size
+            var estimatedStateSize = checked((long)NativeHandle.GetStateSize(sequence));
+
+            // Space for "extra" byte plus a 8 byte header
+            var prefixSize = header.Length + 8;
+
+            // Add enough space for the "extra" data and a 6 byte header
+            var totalFileSize = prefixSize + estimatedStateSize;
+
+            // Map the file and write the bytes directly to it.
+            long writtenBytes = 0;
+            using (var file = MemoryMappedFile.CreateFromFile(filename, FileMode.Create, null, totalFileSize))
+            {
+                using (var view = file.CreateViewAccessor(0, totalFileSize))
+                {
+                    unsafe
+                    {
+                        byte* ptr = null;
+                        view.SafeMemoryMappedViewHandle.AcquirePointer(ref ptr);
+                        try
+                        {
+                            // Write prefix data
+                            BinaryPrimitives.WriteUInt32BigEndian(new Span<byte>(ptr + writtenBytes, 4), FileHeaderMagic);
+                            writtenBytes += 4;
+                            BinaryPrimitives.WriteUInt32BigEndian(new Span<byte>(ptr + writtenBytes, 4), (uint)header.Length);
+                            writtenBytes += 4;
+                            header.CopyTo(new Span<byte>(ptr + writtenBytes, header.Length));
+                            writtenBytes += header.Length;
```

Comment:
All of the other save/load methods are public, it's just this new pair with a header of arbitrary bytes that's not. They could easily be made public if there's a usecase for them.

Reply:
I was thinking that since it's directly applicable to the C# representation of a core llama.cpp component, it would make sense giving direct access to it for users of the nuget package. (e.g. even if they don't use the Batched Executor)

Reply:
Discussed in Discord, we decided to move to a separate file out of the LLamaContext.

ref: https://discord.com/channels/1106946823282761851/1106947264938790972/1231762525738111076